### PR TITLE
permit service open api spec

### PR DESCRIPTION
**Pull request by harshal11869**
This PR contains Open API Spec for S100 Permit Service & S100 Shop Facade Service

### CODE REVIEW

File: S100-Permit-Service-Open-Api-Sepc-v0.1.yaml [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+  - jwtBearerAuth: []
+  
+paths:
+  /permits/{licenceId}:
+    get:
+      description: |
+        This service uses the S-100 Part 15 data protection scheme to generate unsigned PERMIT.XML files for the respective User Permit Numbers (UPNs) for the requested licence and returns the compressed zip file containing all these PERMIT.XML files.
+      operationId: getS100Permits
+      tags:
+        -  "UKHO Only"
```
Comment (vijendr13545) [inline S100-Permit-Service-Open-Api-Sepc-v0.1.yaml:27]: "UKHO Only" tag is not relavant as Permit service is an internal only service.

File: S100-Permit-Service-Open-Api-Sepc-v0.1.yaml [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    description: Proposed URL
+
+security:
+  - jwtBearerAuth: []
+  
+paths:
+  /permits/{licenceId}:
+    get:
+      description: |
+        This service uses the S-100 Part 15 data protection scheme to generate unsigned PERMIT.XML files for the respective User Permit Numbers (UPNs) for the requested licence and returns the compressed zip file containing all these PERMIT.XML files.
```
Comment (vijendr13545) [inline S100-Permit-Service-Open-Api-Sepc-v0.1.yaml:24]: This API will generate `signed` permits, though it is outside of WP scope.

File: S100-Permit-Service-Open-Api-Sepc-v0.1.yaml [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+                type: string
+                format: filename=permits.zip; filename*=UTF-8''Permits.zip
+          content:
+            application/zip:
+              schema:
+                title: S100 Permits ZIP File
+                type: string
+                format: binary
+        "204":
+          description: No Content - There are no S100 permits for the licence (No UPNs or Holdings found for the requested licence from S100 Shop Facade).
```
Comment (vijendr13545) [inline S100-Permit-Service-Open-Api-Sepc-v0.1.yaml:55]:
```suggestion
          description: No Content - There are no S100 permits for the licence.
```
I suggest internal wokings of API shouldn't be included.

File: S100-Permit-Service-Open-Api-Sepc-v0.1.yaml [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+                $ref: '#/components/schemas/errorDescription'
+        "429":
+          description: Too Many Requests.
+          headers:
+            X-Correlation-ID:
+              description: GUID for the request for logging/tracing
+              schema:
+                type: string
+        "500":
+          description: Internal Server Error (reasons could be dependent services returns status code 429/401/403/503/500 or error in processing UPNs / permit key or error in generating permit or error in generating zip file).
```
Comment (vijendr13545) [inline S100-Permit-Service-Open-Api-Sepc-v0.1.yaml:122]:
```suggestion
          description: Internal Server Error.
```
I suggest internal wokings of API shouldn't be included in API spec.

File: S100-Permit-Service-Open-Api-Sepc-v0.1.yaml (hunk 2 above)
Comment (harshal11869) [reply S100-Permit-Service-Open-Api-Sepc-v0.1.yaml]: Updated.

File: S100-Permit-Service-Open-Api-Sepc-v0.1.yaml (hunk 3 above)
Comment (harshal11869) [reply S100-Permit-Service-Open-Api-Sepc-v0.1.yaml]: Agree, Updated.

File: S100-Permit-Service-Open-Api-Sepc-v0.1.yaml (hunk 4 above)
Comment (harshal11869) [reply S100-Permit-Service-Open-Api-Sepc-v0.1.yaml]: Agree, Updated.

File: S100-Permit-Service-Open-Api-Sepc-v0.1.yaml [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+security:
+  - jwtBearerAuth: []
+  
+paths:
+  /permits/{licenceId}:
+    get:
+      description: |
+        It uses the S-100 Part 15 data protection scheme to generate signed PERMIT.XML files for all the User Permit Numbers (UPNs) for the requested licence and returns a compressed zip file containing all these PERMIT.XML files.
+      operationId: getS100Permits
+      tags:
```
Comment (rockydevnet) [inline S100-Permit-Service-Open-Api-Sepc-v0.1.yaml:26]: No tags, so I would suggest deleting lines 26 & 27.

File: S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml [hunk 6]
```diff
@@ -0,0 +1,1145 @@
+openapi: 3.0.1
+info:
+  title: UKHO S100 Shop Facade Service APIs
+  description: |
+    The UKHO S100 Shop Facade Service APIs.
+  contact:
+    name: UKHO
+    email: contact@ukho.co.uk
```
Comment (rockydevnet) [inline S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml:8]: If we're going to have a contact address in here, it should be valid. Suggest:

```suggestion
    email: martin.rock-evans@ukho.co.uk
```

File: S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml [hunk 7]
```diff
@@ -0,0 +1,1145 @@
...
+    description: Proposed Dev URL
+security:
+  - jwtBearerAuth: []
+  
+paths:
+  /permits/S100/{licenceId}:
+    get:
+      description: Returns a zip file containing all the S100 permit file(s) for the requested licence.
+      operationId: getS100Permits
+      tags:
```
Comment (rockydevnet) [inline S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml:21]: Tags are empty, so I would suggest deleting lines 21 & 22

File: S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml [hunk 8]
```diff
@@ -0,0 +1,1145 @@
...
+                type: string
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/exceptionDescription'
+            application/zip:
+              schema:
+                $ref: '#/components/schemas/exceptionDescription'
+  
+  /userpermits/{licenceId}/s100:
```
Comment (rockydevnet) [inline S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml:131]: We should try and be consistent in the paths and have either {licenceId} first, or s100 first, when compared to the `getPermits` operation.  My feeling is that both should have the licenceId first (so this one is correct).

File: S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml [hunk 9]
```diff
@@ -0,0 +1,1145 @@
...
+    email: contact@ukho.co.uk
+  version: 0.0.1
+servers:
+  - url: https://shop-facade-service-url
+    description: Proposed Dev URL
+security:
+  - jwtBearerAuth: []
+  
+paths:
+  /permits/S100/{licenceId}:
```
Comment (rockydevnet) [inline S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml:17]: We should try and be consistent in the paths and have either {licenceId} first, or s100 first, when compared to the `getUserPermits` operation.  My feeling is that both should have the licenceId first.

```suggestion
  /permits/{licenceId}/S100:
```

File: S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml [hunk 10]
```diff
@@ -0,0 +1,1145 @@
...
+                $ref: '#/components/schemas/exceptionDescription'
+            application/zip:
+              schema:
+                $ref: '#/components/schemas/exceptionDescription'
+  
+  /userpermits/{licenceId}/s100:
+    get:
+      description: Returns all User Permit Numbers (UPNs) associated with the requested licence. There can be one or more UPNs for a licence.
+      operationId: getUserPermits
+      tags:
```
Comment (rockydevnet) [inline S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml:135]: remove empty tags?

File: S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml [hunk 11]
```diff
@@ -0,0 +1,1145 @@
...
+            X-Correlation-ID:
+              description: GUID for the request for logging/tracing
+              schema:
+                type: string
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/exceptionDescription'
+                
+  /holdings/{licenceId}/s100:
```
Comment (rockydevnet) [inline S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml:240]: We should try and be consistent in the paths and have either {licenceId} first, or s100 first, when compared to the `getPermits` and `getUserPermits` operation.  My feeling is that both should have the licenceId first (so this one is correct).

File: S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml [hunk 12]
```diff
@@ -0,0 +1,1145 @@
...
+    get:
+      description: Get holdings for the requested licence.
+      operationId: getHoldings
+      tags:
```
Comment (rockydevnet) [inline S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml:244]: remove empty tags?

File: S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml [hunk 13]
```diff
@@ -0,0 +1,1145 @@
...
+          example: 12345678
+      responses:
+        "200":
+          description: OK - Returns holdings available for the licence.
+          content:
+            application/json:
+              schema:
+                type: object
+                properties:
+                  productCode:
```
Comment (rockydevnet) [inline S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml:262]: I think this should be ```unitName```

File: S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml [hunk 14]
```diff
@@ -0,0 +1,1145 @@
...
+                    type: string
+                    description: ProductCode define the type of product.
+                    example: "P1231"
+                  productTitle:
```
Comment (rockydevnet) [inline S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml:266]: This should either be `title` or `unitTitle`.

File: S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml [hunk 15]
```diff
@@ -0,0 +1,1145 @@
...
+                    type: string
+                    description: productTitle define the title of product.
+                    example: "P1231"
+                  expiryDate:
+                    type: string
+                    description: expiryDate of the product.
+                    example: "2026-01-31T23:59:00Z"
+                  cells:
```
Comment (rockydevnet) [inline S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml:274]: In S-100 terminology, I believe this should be `datasets`. Each item in this array is then an individual S-100 dataset (which we call cells in the S-57 world, but I think products in the S-100 world, where cell doesn't make sense in some of the different S-100 types).

File: S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml [hunk 16]
```diff
@@ -0,0 +1,1145 @@
...
+                    type: array
+                    items:
+                      $ref: "#/components/schemas/cells"
```
Comment (rockydevnet) [inline S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml:277]: Unless they are used elsewhere, would it be easier to maintain if the array items are inline?

File: S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml [hunk 17]
```diff
@@ -0,0 +1,1145 @@
...
+          - Error
+          - Created
+          - ExpectsShipCoApproval
+          - ExpectsDistributorApproval
+          - ExpectsUkhoApproval
+          - ExternalProcessing
+          - Completed
+          - Rejected
+          - FulfillmentInProgress
+    cells:
```
Comment (rockydevnet) [inline S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml:1028]: I think in the S-100 world, these are datasets.

```suggestion
    datasets:
```

File: S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml [hunk 18]
```diff
@@ -0,0 +1,1145 @@
...
+      type: object
+      properties:
+        cellCode:
+          type: string
+          example: "101GB40079ABCDEFG.000"
+        cellTitle:
+          type: string
+          example: "101GB40079ABCDEFG"
+        latestEditionNumber:
+          type: string
```
Comment (rockydevnet) [inline S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml:1039]:
```suggestion
          type: number
```

File: S100-Permit-Service-Open-Api-Sepc-v0.1.yaml [hunk 19]
```diff
@@ -0,0 +1,162 @@
+openapi: 3.0.1
+info:
+  title: UKHO S100 Permit Service API
+  description: |
+    The UKHO S100 Permit Service API.
+  contact:
+    name: UKHO
+    email: contact@ukho.co.uk
```
[the remark this answers is not in the document]
Comment (martyn-fewtrell) [reply S100-Permit-Service-Open-Api-Sepc-v0.1.yaml:8]: or perhaps abzudeliveryteam@UKHO.gov.uk which is more generic and is valid?

[the remark this answers is not in the document]
Comment (harshal11869) [reply S100-Permit-Service-Open-Api-Sepc-v0.1.yaml]: Updated.

File: S100-Permit-Service-Open-Api-Sepc-v0.1.yaml (hunk 5 above)
Comment (harshal11869) [reply S100-Permit-Service-Open-Api-Sepc-v0.1.yaml]: Deleted.

File: S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml (hunk 6 above)
Comment (harshal11869) [reply S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml]: Updated.

File: S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml (hunk 8 above)
Comment (harshal11869) [reply S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml:129]: Ok,

File: S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml (hunk 9 above)
Comment (harshal11869) [reply S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml]: Updated.

File: S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml (hunk 11 above)
Comment (harshal11869) [reply S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml:236]: Ok.

File: S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml (hunk 12 above)
Comment (harshal11869) [reply S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml]: Removed.

File: S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml (hunk 16 above)
Comment (harshal11869) [reply S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml]: Updated.

File: S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml (hunk 15 above)
Comment (harshal11869) [reply S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml]: Updated.

File: S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml (hunk 17 above)
Comment (harshal11869) [reply S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml]: Updated.

File: S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml [hunk 20]
```diff
@@ -0,0 +1,1145 @@
...
+          - ExpectsUkhoApproval
+          - ExternalProcessing
+          - Completed
+          - Rejected
+          - FulfillmentInProgress
+    cells:
+      description: Object representing the details of cell.
+      type: object
+      properties:
+        cellCode:
```
[the remark this answers is not in the document]
Comment (harshal11869) [reply S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml]: Updated.

File: S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml [hunk 21]
```diff
@@ -0,0 +1,1145 @@
...
+          type: string
+          example: "101GB40079ABCDEFG.000"
+        cellTitle:
```
[the remark this answers is not in the document]
Comment (harshal11869) [reply S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml]: Updated.

File: S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml (hunk 18 above)
Comment (harshal11869) [reply S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml]: Updated.

File: S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml [hunk 22]
```diff
@@ -0,0 +1,1145 @@
...
+          type: string
+          example: "101GB40079ABCDEFG"
+        latestEditionNumber:
+          type: string
+          example: "1"
+        latestUpdateNumber:
+          type: string
```
[the remark this answers is not in the document]
Comment (harshal11869) [reply S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml]: Updated.

File: S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml (hunk 13 above)
Comment (harshal11869) [reply S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml]: Updated.

File: S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml (hunk 7 above)
Comment (harshal11869) [reply S100-Shop-Facade-Open-Api-Sepc-v0.1.yaml]: Removed.

File: S100-Permit-Service-Open-Api-Sepc-v0.1.yaml [hunk 23]
```diff
@@ -0,0 +1,162 @@
...
+  version: 0.0.1
+  
+externalDocs:
+  url: https://github.com/UKHO/s-100-permit-service/tree/main
```
[the remark this answers is not in the document]
Comment (harshal11869) [reply S100-Permit-Service-Open-Api-Sepc-v0.1.yaml]: Updated.

**Review by jrippington — APPROVED** [review]
Looks good, thanks!

**Pull request merged**